### PR TITLE
Add unstructured serialization test

### PR DESCRIPTION
**Pull request by pmorie**
To help prevent #1259; looks like there are some fields exhibiting weird behaviors.

cc @ash2k @nilebox

### CODE REVIEW

**Comment by nilebox** [comment]
@pmorie The commit message and PR title are misleading. Do you plan to add a test in this PR?

**Comment by pmorie** [comment]
oops

**Comment by pmorie** [comment]
I managed to lose this after forgetting to `git add` it an errant `git clean`

**Comment by pmorie** [comment]
@nilebox @ash2k did the work again, should be ready to review now.  I had to use custom fuzzers for objects with parameters because I didn't get the same serializations between unmarshalling from objects and unstructured.  I suspect we are hitting this first because we are (or appear to be) the first to try this type of test on a kind with RawExtension fields.

**Comment by ash2k** [comment]
LGTM

File: pkg/apis/servicecatalog/unstructured_test.go [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+// 9. does semantic equal on (1) and (8)
+func doUnstructuredRoundTrip(t *testing.T, group testapi.TestGroup, kind string) {
+	// We do fuzzing on the internal version of the object, and only then
+	// convert to the external version. This is because custom fuzzing
+	// function are only supported for internal objects.
+	internalObj, err := api.Scheme.New(group.InternalGroupVersion().WithKind(kind))
+	if err != nil {
+		t.Fatalf("Couldn't create internal object %v: %v", kind, err)
+	}
+	seed := rand.Int63()
```
Comment (vaikas-google) [inline pkg/apis/servicecatalog/unstructured_test.go]: Would it be useful to stash away the used seed or log it for being able to repro test failures?

File: pkg/apis/servicecatalog/unstructured_test.go [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+// 4. marshals the unstructured object in (3)
+// 5. unmarshals (4) back into a versioned type
+// 6. does semantic equal on (5) and (1)
+// 7. uses the default converter to convert (1) directly into a map[string]interface{}
+// 8. converts (7) back into a versioned object
+// 9. does semantic equal on (1) and (8)
+func doUnstructuredRoundTrip(t *testing.T, group testapi.TestGroup, kind string) {
+	// We do fuzzing on the internal version of the object, and only then
+	// convert to the external version. This is because custom fuzzing
+	// function are only supported for internal objects.
```
Comment (vaikas-google) [inline pkg/apis/servicecatalog/unstructured_test.go]: nit: functions

File: pkg/apis/servicecatalog/v1alpha1/types.go [hunk 3]
```diff
@@ -429,7 +429,7 @@ type ServiceInstanceSpec struct {
 	// instance. This field is set by the API server and not settable by the
 	// end-user. User-provided values for this field are not saved.
 	// +optional
-	UserInfo *UserInfo `json:"userInfo"`
+	UserInfo *UserInfo `json:"userInfo,omitempty"`
```
Comment (vaikas-google) [inline pkg/apis/servicecatalog/v1alpha1/types.go]: doesn't this cause generated file changes?

File: pkg/apis/servicecatalog/unstructured_test.go (hunk 1 above)
Comment (pmorie) [reply pkg/apis/servicecatalog/unstructured_test.go]: We don't do that for other tests like this -- and I've never seen a failure in one of these tests that wasn't reproducible across runs that I can remember.

File: pkg/apis/servicecatalog/v1alpha1/types.go (hunk 3 above)
Comment (pmorie) [reply pkg/apis/servicecatalog/v1alpha1/types.go]: I guess I'm about to find out!  Definitely openapi changes

File: pkg/apis/servicecatalog/unstructured_test.go (hunk 1 above)
Comment (ash2k) [reply pkg/apis/servicecatalog/unstructured_test.go]: Actually, there may be an issue here - seed for the rand is not initialized. Here is the godoc from the `rand.Seed()` function:
```
// Seed uses the provided seed value to initialize the default Source to a
// deterministic state. If Seed is not called, the generator behaves as
// if seeded by Seed(1). Seed values that have the same remainder when
// divided by 2^31-1 generate the same pseudo-random sequence.
// Seed, unlike the Rand.Seed method, is safe for concurrent use.
```
That means sequence is the same each run? Well, there may be some non-determinism introduced by something else, but this is not ideal.

This is how RNG is seeded in Smith for example (I don't see anything like that in svc-cat in the package with this test): https://github.com/atlassian/smith/blob/master/it/main_test.go

**Pull request merged**